### ISSUE TITLE
__repr__ and __str__ methods

### ISSUE THREAD

**Issue by janniklasrose** [chebpy/chebpy#46]
According to [this](https://stackoverflow.com/a/2626364)

> - `__repr__` goal is to be unambiguous
> - `__str__` goal is to be readable

Other sources also say that `__repr__` is for developers, `__str__` for users.

The current behaviour seems like the opposite:

```python
c = chebfun(lambda x: x**2)
print('str(c)')
print(str(c))
print('-----------------------')
print('repr(c)')
print(repr(c))
```

prints

```
str(c)
<chebfun-col,1,3>
-----------------------
repr(c)
chebfun column (1 smooth piece)
       interval       length     endpoint values
[      -1,       1]        3         1        1
vertical scale =   1
```

Note also that any user will probably want to call `print(c)` by default and expect to get the second output, i.e. the `__repr__` description.

**Comment by markrichardson** [comment]
I like the current accessibility of

```
chebfun column (1 smooth piece)
       interval       length     endpoint values
[      -1,       1]        3         1        1
vertical scale =   1
```

a lot and I think it's important to have this displayed by default for easy inspection of the chebfun (say in a terminal or a notebook). Thus not sure I like the idea of forcing the user to `print(c)` in order to get this information (if that's what the proposal is? I may have misunderstood.)

Both numpy and pandas have inspectable default output under the same conditions, so this is the precedent I would refer to in Python world to justify what is currently happening in chebpy is not that weird or unconventional:

```
np.random.randn(3, 4)
Out[9]: 
array([[-0.01307649,  0.50097916,  1.159666  ,  0.59126538],
       [-0.94939886,  0.24740005, -1.03903929,  1.29590136],
       [ 1.64651049,  2.83397725, -0.61213611, -0.45840565]])
```

```
pd.DataFrame(np.random.randn(3, 4))
Out[8]: 
          0         1         2         3
0 -1.980832  0.098227  2.099621  0.321122
1 -0.933835  3.214470 -1.089083  1.192333
2 -0.836502  0.260645  0.801251 -0.374539
```

As an aside I'm not a fan of current str output `<chebfun-col,1,3>` and would be quite happy to change it to something better.

**Comment by janniklasrose** [comment]
I agree that the default output in a terminal/iPython session should be straightforward and verbose, as it is for numpy and pandas. What I'm suggesting, and I think you agree, is to improve the `print` (`__str__`) implementation. Both numpy and pandas have (almost) the same behaviour with `__str__` and `__repr__`, though, which is where something should be done.

My initial suggestion was more to increase verbosity in both, to make `__str__` essentially equivalent to the current (pretty) `__repr__` and add more information to `__repr__` (if there even is anything missing?). I think it would be sufficient for now to set `__str__`=`__repr__` (or remove `__str__` entirely, which causes Python to default to this equality).

**Comment by markrichardson** [comment]
As I remember the original logic for having `__str__` != `__repr__` primarily applied to the lower level classes (`Bndfun`, `ChebTech2` etc). The purpose was to be able to do easy inspection of funs stored in a numpy array, for example:

```
from chebpy import pwc

f = pwc()

f.funs
Out[23]: array([Bndfun([-1.0, 0.0], 1), Bndfun([0.0, 1.0], 1)], dtype=object)
``` 

(As numpy repr calls `__str__` on the components.)

This is less important for top-level `Chebfun` so we can potentially kill `__str__` (in `chebpy.core.chebfun/Chebfun`) as suggested. 

But to be clear on the goal,  I get the same visual output here:

```
f
Out[25]: 
chebfun column (2 smooth pieces)
       interval       length     endpoint values
[      -1,       0]        1         0        0
[       0,       1]        1         1        1
vertical scale =   1    total length = 2

print(f)
chebfun column (2 smooth pieces)
       interval       length     endpoint values
[      -1,       0]        1         0        0
[       0,       1]        1         1        1
vertical scale =   1    total length = 2
```
So apparent no difference between `f [enter]` and `print(f)`? Or have I missed the point?

**Comment by janniklasrose** [comment]
That's odd, in my iPython I get a different behaviour:

```
In [1]: from chebpy import chebfun

In [2]: f = chebfun(lambda x: x**2)

In [3]: f
Out[3]:
chebfun column (1 smooth piece)
       interval       length     endpoint values
[      -1,       1]        3         1        1
vertical scale =   1

In [4]: print(f)
<chebfun-col,1,3>


In [5]: str(f)
Out[5]: '<chebfun-col,1,3>\n'

In [6]: repr(f)
Out[6]: 'chebfun column (1 smooth piece)\n       interval       length     endpoint values\n[      -1,       1]        3         1        1\nvertical scale =   1'

In [7]: print(repr(f))
chebfun column (1 smooth piece)
       interval       length     endpoint values
[      -1,       1]        3         1        1
vertical scale =   1
```

That's what [the documentation](https://docs.python.org/3/library/functions.html#print) says should happen:

> All non-keyword arguments are converted to strings like str() does [...]

**Comment by janniklasrose** [comment]
> As I remember the original logic for having __str__ != __repr__ primarily applied to the lower level classes (Bndfun, ChebTech2 etc). The purpose was to be able to do easy inspection of funs stored in a numpy array, for example:

I agree, and those seem to use the `__repr__` representation.

Here a quick test:

```python
class Func_str_only:
    def __str__(self):
        return "Func_str"
class Func_repr_only:
    def __repr__(self):
        return "Func_repr"
class Func_both:
    def __str__(self):
        return "Func_str"
    def __repr__(self):
        return "Func_repr"
```

| Definition | `f [ENTER]` | `str(f)` | `repr(f)` | `print(f)` | `[f]` |
| ----------- | ----------- | ------- | -------- | --------- | ---- |
| `Func_str_only` | `<__main__.Func_str_only at 0x7f196464d050>` | `'Func_str'` | `'<__main__.Func_str_only object at 0x7f196464d050>'` | `Func_str` | `['<__main__.Func_str_only object at 0x7f196464d050>']` |
| `Func_repr_only` | `Func_repr` | `'Func_repr'` | `'Func_repr'` | `Func_repr` | `[Func_repr]` |
| `Func_both` | `Func_repr` | `'Func_str'` | `'Func_repr'` | `'Func_str'` | `[Func_repr]` |

I think what we want is
- `Func_repr_only` (only define `__repr__`) for low-level functions
- `Func_both` (`__str__ != __repr__`) for high-level (user) functions like `chebfun`, with the following change:
  - `__str__` is the verbose output currently seen from `__repr__`, so that `print(f)` does what you would expect
  - `__repr__` is a brief one-liner similar to the current `__str__`, so that using `chebfun` in a list is readable (currently you get an ugly multi-line mess).

This has the caveat that a user typing `f [ENTER]` in the terminal will only see the `__repr__` version (which we could make a bit nicer than the brief version in the current `chebfun` class), but I personally think `f [ENTER]` is not very pythonic compared to `print(f)` (especially if in a jupyter/iPython cell with multiple outputs)
